packages/components/tree-v2/src/hooks/useTree.ts

```diff
@@ -0,0 +1,285 @@
+  })
+  const disabledKey = computed(() => {
+    return (props.props && props.props.disabled) || 'disabled'
+  })
+  const labelKey = computed(() => {
+    return (props.props && props.props.label) || 'label'
+  })
+
+  const flattenTree = computed(() => {
+    const expandedKeys = expandedKeySet.value
+    const hiddenKeys = hiddenNodeKeySet.value
+    const flattenNodes: TreeNode[] = []
+    const nodes = (tree.value && tree.value.treeNodes) || []
+    function traverse() {
+      const stack: TreeNode[] = []
+      for (let i = nodes.length - 1; i >= 0; --i) {
+        stack.push(nodes[i])
+      }
+      while (stack.length) {
+        const node = stack.pop()
+        if (!hiddenKeys.has(node.key)) {
+          flattenNodes.push(node)
+        }
+        // Only "visible" nodes will be rendered
+        if (expandedKeys.has(node.key)) {
+          const children = node.children
+          if (children) {
+            const length = children.length
+            for (let i = length - 1; i >= 0; --i) {
+              stack.push(children[i])
+            }
+          }
+        }
+      }
+    }
+    traverse()
+    return flattenNodes
+  })
+
+  const isNotEmpty = computed(() => {
+    return flattenTree.value.length > 0
+  })
+
+  function createTree(data: TreeData): Tree {
+    const treeNodeMap: Map<TreeKey, TreeNode> = new Map()
+    const levelTreeNodeMap: Map<number, TreeNode[]> = new Map()
+    let maxLevel = 1
+    function traverse(
+      nodes: TreeData,
+      level = 1,
+      parent: TreeNode | null = null
+    ) {
+      const siblings: TreeNode[] = []
+      for (let index = 0; index < nodes.length; ++index) {
+        const rawNode = nodes[index]
+        const node: TreeNode = {}
+        node.label = getLabel(rawNode)
+        node.level = level
+        node.parent = parent
+        node.data = rawNode
+        const value = getKey(rawNode)
+        node.key = value
+        const children = getChildren(rawNode)
+        node.disabled = getDisabled(rawNode)
+        node.isLeaf = !children || children.length === 0
+        if (children && children.length) {
+          node.children = traverse(children, level + 1, node)
+        }
+        siblings.push(node)
+        treeNodeMap.set(value, node)
+        if (!levelTreeNodeMap.has(level)) {
+          levelTreeNodeMap.set(level, [])
+        }
+        levelTreeNodeMap.get(level).push(node)
+      }
+      if (level > maxLevel) {
+        maxLevel = level
+      }
+      return siblings
+    }
+    const treeNodes: TreeNode[] = traverse(data)
+    return {
+      treeNodeMap,
+      levelTreeNodeMap,
+      maxLevel,
+      treeNodes,
+    }
+  }
+
+  function filter(query: string) {
+    const keys = doFilter(query)
+    if (keys) {
+      expandedKeySet.value = keys
+    }
+  }
+
+  function getChildren(node: TreeNodeData): TreeNodeData[] {
+    return node[childrenKey.value]
+  }
+
+  function getKey(node: TreeNodeData): TreeKey {
```

Comment:
The return type of this function seems incorrect since if the `node` passed into this function it will return `undefined` instead of `TreeKey`, maybe you should mark it explicitly.